### ISSUE TITLE
Behaviour of intersection with Any

### ISSUE THREAD

**Issue by JanEricNitschke** [astral-sh/ty#4447]
### Question

Not sure if this is a bug so i am asking it as a question.

In this scenario:

```py
from typing import Any, reveal_type

def get() -> Any:
    return 42  # pretend this could be anything

x = get()
reveal_type(x)

y = get()
assert isinstance(y, int)
reveal_type(y)  # `int` (mypy/pyright) / `Any & int` (ty)

print(y.bit_count())
print(y.blablabla())

```

Mypy, pyright and pyrefly give me the revealed types as `Any` and `int` and complain on the last line.

However ty gives me `Any & int` and does not complain.

Not sure if that is working as intended or if there is a bug here.

[Pyrefly playground](https://pyrefly.org/sandbox/?project=v2.lVbLauMwFP0VQWBIF7JhmFVhYLazGOgHzCKKrNiitmz0yMTz9T1Xkh05TVPaNtDcl450zz3SVyWzFBvqOD7U7Vb51GwEPZOxINCP74zt0BzllWmY70BQOYa-YUfFhMGZYY2l1gWbjLXoS7HW_hIt9JnLkCSMJZeBESx4lz8_EYgDXAe2H-ZprtE83XaYypodgJp9ozw4_bwulQg1V0ecLCDj_6foW-0kpvQH-3s67NhLMrB8stg2eNQGyC34Uf01OyitYp33k3uuCVBMH21bK1M3o3T1JqNmxHxxFjqK-LYcGyN9Hco-4CG64JSPJkw7LhwyanNSlv_TvuMnbZ3nAWP9k3nMDo3blbsgQq-l9lwckQ1O8aSoVC4GUbVBO3Sj5W4MViZqr75HzF6DxkkZHswk5Cuxb-MLRlARQeOT6bXxNwo7lOTe2lfkmcs3qAxXl04ER0PCB-Flt40A3GAaYbBhXL4f-kbT6NiLG8xG0S6FnREyTMJq9yiG-Eq1yv4VgZAx2cX-bxw7VrjSHgjsTRTFuLE_xyDCjcZD126DCEJz96zOQE9zxtHmOweVGG0VJeIlA0nSotf_VfM7zyeeOzqCh3bnCbx2p585xEQbrC086HkMfquNGDLnw-n0oUZdn10Wu9C4TmSn5CutuAz1r7ueROQXYZ36MzahV_ulFK7ErGfYIH4XexxGGlIABQC0FYc9xFwMYQxdllxeBjEmPwy8uvj4XMkXe1RPfF1Xo5-qqkqVTfmfZubBoDgGjITk4R3wBg)

[Pyright playground]()

[Mypy playground](https://mypy-play.net/?gist=611bf1917ef13b080f42f2f162fa3b62)

[ty playground](https://play.ty.dev/33a87d65-bc99-42b4-9cb4-3af1baec33fe)

**Comment by sharkdp** [comment]
> However ty gives me `Any & int` and does not complain.
> 
> Not sure if that is working as intended or if there is a bug here.

This is working as intended and not a bug. `get()` could return an object of a class that inherits from `int` AND implements the `blablabla` method, and so there is a possible interpretation of that program where that `y.blablabla()` call is not an error.

`Any & int` is a gradual type that could materialize to `int` *or to any subtype of `int`*. This type can not be represented in those other type checkers, as far as I know, which is why they need to treat this situation in a less permissive way (or choose `Any`, which would be *too* permissive). 

If you don't want the permissive behavior, you should consider giving your `get()` function a return type of `object` instead.
